File: sem4/mpp/project-java/repository/src/main/java/repository/SignupDBRepository.java
package repository;

import domain.AgeGroup;
import domain.RaceDistance;
import domain.Signup;
import domain.Child;
import utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;

public class SignupDBRepository implements SignupRepository {
	private final JdbcUtils dbUtils;
	private final ChildRepository childrepository;
	private static final Logger logger = LogManager.getLogger(SignupDBRepository.class.getName());

	public SignupDBRepository(Properties props) {
		logger.info("Initializing user repository with properties: {} ", props);
		dbUtils = new JdbcUtils(props);
		childrepository = new ChildDBRepository(props);
	}

	@Override
	public Signup add(Signup elem) {
		logger.traceEntry("add signup {}", elem);
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("INSERT INTO Signup (child_id, age_group, race_distance) values (?,?,?)")) {
			preparedStatement.setInt(1, elem.getChild().getId());
			preparedStatement.setString(2, elem.getAgeGroup().toString());
			preparedStatement.setString(3, elem.getRaceDistance().toString());
			int result = preparedStatement.executeUpdate();
			logger.trace("saved {}", result);
			ResultSet resultSet = con.prepareStatement("SELECT last_insert_rowid()").executeQuery();
			resultSet.next();
			elem.setId(resultSet.getInt(1));
			logger.trace("set id {}", elem.getId());
		} catch (SQLException e) {
			logger.error(e);
			throw new RuntimeException(e);
		}
		logger.traceExit();
		return elem;
	}

	@Override
	public void delete(Integer id) {
		logger.traceEntry("delete signup");
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("DELETE FROM Signup WHERE id=?")) {
			preparedStatement.setInt(1, id);
			int result = preparedStatement.executeUpdate();
			logger.trace("deleted {}", result);
		} catch (SQLException e) {
			logger.error(e);
			throw new RuntimeException(e);
		}
		logger.traceExit();
	}

	@Override
	public void update(Signup elem, Integer id) {
		logger.traceEntry("update signup {}", elem);
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("UPDATE Signup SET child_id=?, age_group=?, race_distance=? WHERE id=?")) {
			preparedStatement.setInt(1, elem.getChild().getId());
			preparedStatement.setString(2, elem.getAgeGroup().toString());
			preparedStatement.setString(3, elem.getRaceDistance().toString());
			preparedStatement.setInt(4, id);
			int result = preparedStatement.executeUpdate();
			logger.trace("saved {}", result);
		} catch (SQLException e) {
			logger.error(e);
			throw new RuntimeException(e);
		}
		logger.traceExit();
	}

	@Override
	public Signup getById(Integer id) {
		logger.traceEntry("get signups {}");
		Connection con = dbUtils.getConnection();
		Signup signup = null;
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Signup WHERE id=?")) {
			preparedStatement.setInt(1, id);
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				if (resultset.next()) {
					Child child = childrepository.getById(resultset.getInt("child_id"));
					AgeGroup ageGroup = AgeGroup.valueOf(resultset.getString("age_group"));
					RaceDistance raceDistance = RaceDistance.valueOf(resultset.getString("race_distance"));
					signup = new Signup(child, ageGroup, raceDistance);
					signup.setId(id);
				} else {
					logger.error("no signup with id");
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return signup;
	}

	@Override
	public Collection<Signup> getAll() {
		logger.traceEntry("get signups {}");
		Connection con = dbUtils.getConnection();
		List<Signup> signups = new ArrayList<>();
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Signup")) {
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				while (resultset.next()) {
					Integer id = resultset.getInt("id");
					Child child = childrepository.getById(resultset.getInt("child_id"));
					AgeGroup ageGroup = AgeGroup.valueOf(resultset.getString("age_group"));
					RaceDistance raceDistance = RaceDistance.valueOf(resultset.getString("race_distance"));
					Signup signup = new Signup(child, ageGroup, raceDistance);
					signup.setId(id);
					signups.add(signup);
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return signups;
	}

	@Override
	public Collection<Child> getChildrenByRace(AgeGroup ageGroup, RaceDistance raceDistance) {
		logger.traceEntry("get signups {}");
		Connection con = dbUtils.getConnection();
		List<Child> children = new ArrayList<>();
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Signup WHERE age_group=? AND race_distance=?")) {
			preparedStatement.setString(1, ageGroup.toString());
			preparedStatement.setString(2, raceDistance.toString());

			try (ResultSet resultset = preparedStatement.executeQuery()) {
				while (resultset.next()) {
					Child child = childrepository.getById(resultset.getInt("child_id"));
					children.add(child);
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return children;
	}

	@Override
	public Integer getChildSignups(Integer id) {
		logger.traceEntry("get no of signups {}");
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT COUNT(*) FROM Signup WHERE child_id=?")) {
			preparedStatement.setInt(1, id);
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				if (resultset.next()) {
					return resultset.getInt(1);
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return 0;
	}
}

File: sem4/mpp/project-cs/repository/SignupDBRepository.cs
using project_cs.domain;

namespace project_cs.repo;

public class SignupDbRepository : ISignupRepository {
    private static readonly log4net.ILog Log =
        log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

    private readonly IChildRepository _childRepository;
    readonly IDictionary<String, string> _props;

    public SignupDbRepository(IDictionary<string, string> props) {
        Log.Info("creating signup repo ");
        this._props = props;
        _childRepository = new ChildDbRepository(props);
    }

    public Signup Add(Signup elem) {
        Log.Info("add signup");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText =
                "INSERT INTO Signup (child_id, age_group, race_distance) VALUES (@child_id, @age_group, @race_distance)";

            var paramChild = comm.CreateParameter();
            paramChild.ParameterName = "@child_id";
            paramChild.Value = elem.child.id;
            comm.Parameters.Add(paramChild);

            var paramAge = comm.CreateParameter();
            paramAge.ParameterName = "@age_group";
            paramAge.Value = elem.ageGroup.ToString();
            comm.Parameters.Add(paramAge);

            var paramDistance = comm.CreateParameter();
            paramDistance.ParameterName = "@race_distance";
            paramDistance.Value = elem.raceDistance.ToString();
            comm.Parameters.Add(paramDistance);
            comm.ExecuteNonQuery();

            var comm2 = con.CreateCommand();
            comm2.CommandText = "SELECT last_insert_rowid()";
            var id = (Int64)comm2.ExecuteScalar()!;
            elem.id = (int)id;
            return elem;
        }
    }

    public void Delete(int id) {
        Log.Info("delete signup");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "DELETE FROM Signup WHERE ID=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);
            comm.ExecuteNonQuery();
        }
    }

    public void Update(Signup elem, int id) {
        Log.Info("update signup");
        var con = DbUtils.GetConnection(_props);
        using var comm = con.CreateCommand();
        comm.CommandText =
            "UPDATE Signup SET child_id=@child_id, age_group=@age_group, race_distance=@race_distance WHERE id=@id";
        var paramId = comm.CreateParameter();
        paramId.ParameterName = "@id";
        paramId.Value = id;
        comm.Parameters.Add(paramId);

        var paramChild = comm.CreateParameter();
        paramChild.ParameterName = "@child_id";
        paramChild.Value = elem.child.id;
        comm.Parameters.Add(paramChild);

        var paramAge = comm.CreateParameter();
        paramAge.ParameterName = "@age_group";
        paramAge.Value = elem.ageGroup.ToString();
        comm.Parameters.Add(paramAge);

        var paramDistance = comm.CreateParameter();
        paramDistance.ParameterName = "@race_distance";
        paramDistance.Value = elem.raceDistance.ToString();
        comm.Parameters.Add(paramDistance);

        comm.ExecuteNonQuery();
    }

    public Signup? GetById(int id) {
        Log.Info("get signup");
        var con = DbUtils.GetConnection(_props);
        Signup? s = null;
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT child_id, age_group, race_distance, id FROM Signup WHERE id=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);
            using (var dataR = comm.ExecuteReader()) {
                if (dataR.Read()) {
                    var child = _childRepository.GetById(dataR.GetInt32(0))!;
                    Enum.TryParse(dataR.GetString(1), out AgeGroup ageGroup);
                    Enum.TryParse(dataR.GetString(2), out RaceDistance raceDistance);
                    s = new Signup(child, ageGroup, raceDistance);
                    s.id = id;
                }
            }
        }

        return s;
    }

    public ICollection<Signup> GetAll() {
        Log.Info("get all from signup");
        var con = DbUtils.GetConnection(_props);
        IList<Signup> signups = new List<Signup>();
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT child_id, age_group, race_distance, id FROM Signup";

            using (var dataR = comm.ExecuteReader()) {
                while (dataR.Read()) {
                    var child = _childRepository.GetById(dataR.GetInt32(0))!;
                    Enum.TryParse(dataR.GetString(1), out AgeGroup ageGroup);
                    Enum.TryParse(dataR.GetString(2), out RaceDistance raceDistance);
                    var s = new Signup(child, ageGroup, raceDistance);
                    s.id = dataR.GetInt32(3);
                    signups.Add(s);
                }
            }
        }

        return signups;
    }

    public ICollection<Child> GetChildrenByRace(AgeGroup age, RaceDistance distance) {
        Log.Info("get children by race");
        var con = DbUtils.GetConnection(_props);
        IList<Child> children = new List<Child>();
        using (var comm = con.CreateCommand()) {
            comm.CommandText =
                "SELECT child_id, age_group, race_distance, id FROM Signup WHERE age_group=@age_group AND race_distance=@race_distance";
            var paramAgeGroup = comm.CreateParameter();
            paramAgeGroup.ParameterName = "@age_group";
            paramAgeGroup.Value = age.ToString();
            comm.Parameters.Add(paramAgeGroup);
            var paramRaceDistance = comm.CreateParameter();
            paramRaceDistance.ParameterName = "@race_distance";
            paramRaceDistance.Value = distance.ToString();
            comm.Parameters.Add(paramRaceDistance);
            using (var dataR = comm.ExecuteReader()) {
                while (dataR.Read()) {
                    var c = _childRepository.GetById(dataR.GetInt32(0))!;
                    children.Add(c);
                }
            }
        }

        return children;
    }

    public int GetChildSignups(int id) {
        Log.Info("get no of signups");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT COUNT(*) FROM Signup WHERE child_id=@child_id";
            var paramChildId = comm.CreateParameter();
            paramChildId.ParameterName = "@child_id";
            paramChildId.Value = id;
            comm.Parameters.Add(paramChildId);
            var dataR = comm.ExecuteScalar();
            return Convert.ToInt32(dataR!.ToString());
        }
    }
}